### PR TITLE
Prepend `DAG:` to dag permissions

### PR DESCRIPTION
**Pull request by jhtimmins**
This adds the prefix `DAG:` to newly created dag permissions. It supports checking permissions on both prefixed and un-prefixed DAG permission names.

This will make it easier to identify permissions that related to granular dag access.

This PR does not modify existing dag permission names to use the new prefixed naming scheme. That will come in a separate PR.

Related to issue #10469

### CODE REVIEW

File: airflow/www/security.py [hunk 1]
```diff
@@ -320,33 +317,75 @@ def get_accessible_dags(self, user_actions, user, session=None):
...
+                if action not in user_actions:
+                    continue
+
+                if resource.startswith(DAG_PREFIX):
+                    resources.add(resource[len(DAG_PREFIX) :])
+                else:
                     resources.add(resource)
 
-        if bool({'Dag', 'all_dags'}.intersection(resources)):
+        if 'Dag' in resources:
```
Comment (ashb) [inline airflow/www/security.py]: I think this is my only concern -- it's not clear from the permission name that "can read on Dag" (as this would be shown in the FAB Permissions screens) would be _all_ dags.

Hmmm.

Comment (ashb) [reply airflow/www/security.py]: I think this should _possibly_ be the exception to the resource naming convention, and call this "All Dags"

File: airflow/www/security.py [hunk 2]
```diff
@@ -320,33 +317,75 @@ def get_accessible_dags(self, user_actions, user, session=None):
...
+            user = g.user
+        prefixed_dag_id = self.prefixed_dag_id(dag_id)
+
+        return self._has_view_access(user, CAN_EDIT, 'Dag') or self._has_view_access(
+            user, CAN_EDIT, prefixed_dag_id
+        )
+
+    def prefixed_dag_id(self, dag_id):
+        """Returns the permission name for a DAG id."""
+        if dag_id == 'Dag':
```
Comment (ashb) [inline airflow/www/security.py]: WDYT about extracting this to a constant to make it clear the significance of this "Dag" string?

Comment (ashb) [reply airflow/www/security.py]: Maybe, not sure.

File: airflow/www/security.py [hunk 3]
```diff
@@ -520,18 +560,23 @@ def merge_pv(perm, view_menu):
...
         """
-        all_dag_view = self.find_view_menu('all_dags')
-        dag_perm_ids = [self.find_permission('can_dag_edit').id, self.find_permission('can_dag_read').id]
+        all_dag_view = self.find_view_menu('Dag')
+        pvs = (
+            self.get_session.query(sqla_models.ViewMenu)
+            .filter(sqla_models.ViewMenu.name.like(f"{DAG_PREFIX}%"))
+            .all()
+        )
+        pv_ids = [pv.id for pv in pvs]
```
Comment (ashb) [inline airflow/www/security.py]:
```suggestion
        pv_ids = (
            self.get_session.query(sqla_models.ViewMenu.id)
            .filter(sqla_models.ViewMenu.name.like(f"{DAG_PREFIX}%"))
            .all()
        )
```
I think?

(Though that might yield `[[id1],[id2]]`)

File: airflow/www/security.py [hunk 4]
```diff
@@ -564,7 +609,7 @@ def sync_roles(self):
             perms = config['perms']
             self.init_role(role, vms, perms)
         self.create_custom_dag_permission_view()
-
+        self.log.warning('Loading sync_roles.')
```
Comment (ashb) [inline airflow/www/security.py]: Debug I assume?

File: airflow/cli/cli_parser.py [hunk 5]
```diff
@@ -1228,6 +1228,12 @@ class GroupCommand(NamedTuple):
         func=lazy_load_command('airflow.cli.commands.role_command.roles_create'),
         args=(ARG_ROLES,),
     ),
+    ActionCommand(
+        name='upgrade',
+        help='Upgrade role permissions',
+        func=lazy_load_command('airflow.cli.commands.role_command.roles_upgrade'),
```
Comment (ashb) [inline airflow/cli/cli_parser.py]: I don't see this function added (yet?)

But maybe we don't need this with the it being done in a migration.

File: airflow/www/security.py (hunk 1 above)
Comment (jhtimmins) [reply airflow/www/security.py]: I think that's a good idea. While it is an exception, it does remove some ambiguity to clarify the name.

File: airflow/www/security.py (hunk 3 above)
Comment (jhtimmins) [reply airflow/www/security.py]: This yields `((id1,), (id2,), (id3,))`

File: airflow/cli/cli_parser.py (hunk 5 above)
Comment (jhtimmins) [reply airflow/cli/cli_parser.py]: Yup. This should be deleted.

File: airflow/www/security.py (hunk 3 above)
Comment (ashb) [reply airflow/www/security.py]: (not saying this is a good idea, or even clear to read) but:

```python shell
>>> (x,) = zip(*((1,),(2,)))                                                                                                                                                                                                                                                                                                                                            
>>> x                                                                                                                                                                                                                                                                                                                                                                   
(1, 2)
```

Comment (jhtimmins) [reply airflow/www/security.py]: Interesting. Good to have this in my back pocket. Although I agree that for now, it probably reduces readability.

**Comment by jhtimmins** [comment]
@ashb @kaxil If y'all can take a look. As long as tests are passing, you can merge if you don't have any issues.

The new support for multiple schedulers seems to cause variance in the number of queries at times, so I added `margin_of_error` to the `assert_queries_count` function. The default is 0, but for some `test_scheduler_job.py` I set it to `2`. So if the expected number of queries is 95, tests will pass with 93-97 queries. This should solve variance in number of queries due to test/DB flakiness.

File: airflow/www/security.py [hunk 6]
```diff
@@ -520,18 +550,23 @@ def merge_pv(perm, view_menu):
...
-        because Admin have already have all_dags permission.
+        because Admin have already have Dag permission.
         Add the missing ones to the table for admin.
 
         :return: None.
         """
-        all_dag_view = self.find_view_menu('all_dags')
-        dag_perm_ids = [self.find_permission('can_dag_edit').id, self.find_permission('can_dag_read').id]
+        all_dag_view = self.find_view_menu(permissions.RESOURCE_ALL_DAGS)
+        pvs = (
```
Comment (ashb) [inline airflow/www/security.py]:
```suggestion
        dag_pvs = (
```

File: tests/test_utils/asserts.py [hunk 7]
```diff
@@ -69,16 +72,22 @@ def after_cursor_execute(self, *args, **kwargs):
 
 
 @contextmanager
-def assert_queries_count(expected_count, message_fmt=None):
+def assert_queries_count(expected_count, message_fmt=None, margin_of_error=0):
```
Comment (ashb) [inline tests/test_utils/asserts.py]: Now that we've quarantined the failing tests (for now) could you revert the change to this file please?

File: tests/jobs/test_scheduler_job.py [hunk 8]
```diff
@@ -3595,7 +3595,7 @@ def test_execute_queries_count_with_harvested_dags(self, expected_query_count, d
             job.heartbeat = mock.MagicMock()
             job.processor_agent = mock_agent
 
-            with assert_queries_count(expected_query_count):
+            with assert_queries_count(expected_query_count, margin_of_error=5):
```
Comment (ashb) [inline tests/jobs/test_scheduler_job.py]: Here too - revert the changes to this file please.

File: airflow/www/security.py [hunk 9]
```diff
@@ -377,13 +406,13 @@ def has_all_dags_access(self):
         """
         Has all the dag access in any of the 3 cases:
         1. Role needs to be in (Admin, Viewer, User, Op).
-        2. Has can_dag_read permission on all_dags view.
-        3. Has can_dag_edit permission on all_dags view.
+        2. Has can_read permission on dags view.
+        3. Has can_edit permission on dags view.
```
Comment (ashb) [inline airflow/www/security.py:418]:
```suggestion
        2. Has can_read permission on AllDags view.
        3. Has can_edit permission on AllDags view.
```

(Not as part of this PR, but why on earth do we check the role membership at all? We should just use the permissions system!)

File: airflow/api_connexion/security.py [hunk 10]
```diff
@@ -54,7 +55,7 @@ def check_authorization(
         return
     appbuilder = current_app.appbuilder
     for permission in permissions:
-        if permission in (('can_read', 'Dag'), ('can_edit', 'Dag')):
+        if permission in (('can_read', RESOURCE_ALL_DAGS), ('can_edit', RESOURCE_ALL_DAGS)):
```
Comment (ashb) [inline airflow/api_connexion/security.py]: I'm not sure about this one -- the permission we declare on the models feels like it should still be `Dag` -- otherwise it reads as if we are asking explicitly for permission on AllDags, not just "the requested Dag".

File: airflow/cli/commands/role_command.py [hunk 11]
```diff
@@ -29,9 +29,7 @@ def roles_list(args):
     roles = appbuilder.sm.get_all_roles()
     print("Existing roles:\n")
     role_names = sorted([[r.name] for r in roles])
-    msg = tabulate(role_names,
-                   headers=['Role'],
-                   tablefmt=args.output)
+    msg = tabulate(role_names, headers=['Role'], tablefmt=args.output)
```
Comment (ashb) [inline airflow/cli/commands/role_command.py]: Could you undo this please? (We like to not have whitespace-only/style changes in files otherwise untouched in a PR)

File: airflow/models/dag.py [hunk 12]
```diff
@@ -176,7 +176,7 @@ class DAG(BaseDag, LoggingMixin):
         that it is executed when the dag succeeds.
     :type on_success_callback: callable
     :param access_control: Specify optional DAG-level permissions, e.g.,
-        "{'role1': {'can_dag_read'}, 'role2': {'can_dag_read', 'can_dag_edit'}}"
+        "{'role1': {'can_read'}, 'role2': {'can_read', 'can_edit'}}"
```
Comment (ashb) [inline airflow/models/dag.py:180]: I think somewhere we should have an "upgrade" of the old style of permissions names for this param -- otherwise when upgrading custom DAG permissions would stop working and not give any obvoius reason why.

File: airflow/www/security.py [hunk 13]
```diff
@@ -320,33 +307,75 @@ def get_accessible_dags(self, user_actions, user, session=None):
...
-            return self.is_item_public(permission, view_name)
-        return self._has_view_access(user, permission, view_name)
+            return self.is_item_public(permission, resource)
+
+        has_access = self._has_view_access(user, permission, resource)
+
+        if permission == permissions.ACTION_CAN_READ:
+            has_access |= self.can_read_dag(resource, user)
+        elif permission == permissions.ACTION_CAN_EDIT:
+            has_access |= self.can_edit_dag(resource, user)
```
Comment (ashb) [inline airflow/www/security.py]: I think this is going to return true too often:

For example, if the permission check is `can_edit on Variables`, then can_edit_dag checks for the following:

- "can_edit on AllDags" or
- "can_edit on DAG:Variables"

This means the if anyone has can_edit on AllDags, they get can_edit on _everything_.

We need to improve the tests as this wasn't caught by them and so needs to be.

File: airflow/www/views.py [hunk 14]
```diff
@@ -545,7 +546,7 @@ def dag_stats(self, session=None):
         dr = models.DagRun
 
         allowed_dag_ids = current_app.appbuilder.sm.get_accessible_dag_ids(g.user)
-        if 'all_dags' in allowed_dag_ids:
+        if permissions.RESOURCE_ALL_DAGS in allowed_dag_ids:
```
Comment (ashb) [inline airflow/www/views.py]:
```suggestion
        if permissions.RESOURCE_ALL_DAGS in allowed_dag_ids:
```

This won't ever be true anymore will it, cos `get_accessible_dag_ids` already handles this an filters it to return all dag ids. (This is probably a separate PR, and this change is fine here)

**Review by ashb — CHANGES_REQUESTED** [review]
Some changes, and I think there's a bug in here not caught by the tests, so we need to expand the tests

File: airflow/www/views.py (hunk 14 above)
Comment (jhtimmins) [reply airflow/www/views.py]: Agreed, and agreed.

File: airflow/www/security.py (hunk 9 above)
Comment (jhtimmins) [reply airflow/www/security.py:418]: Yeah permissions should be refactored/simplified in general, but I've been avoiding that as much as possible.

File: airflow/models/dag.py (hunk 12 above)
Comment (jhtimmins) [reply airflow/models/dag.py:180]: I'm not sure what you mean by this?

File: airflow/www/security.py (hunk 13 above)
Comment (jhtimmins) [reply airflow/www/security.py]: Shoot, you're right. Good catch. I'll add a test to `test_security.py`.

File: airflow/api_connexion/security.py (hunk 10 above)
Comment (jhtimmins) [reply airflow/api_connexion/security.py]: So you're suggesting that the backend permission be `AllDags`, but on the views its `Dag`? That feels dicey. How would you feel about access to `Dags` as resource name? Allows the possibility of multiple Dags, but without overly emphasizing the `All` aspect of it?

File: airflow/models/dag.py (hunk 12 above)
Comment (ashb) [reply airflow/models/dag.py:180]: If I have this in my dag file:

```
with DAG(..., access_control={'role1': {'can_dag_read'}, 'role2': {'can_dag_read', 'can_dag_edit'}})
```

What happens to the permissions.

Comment (jhtimmins) [reply airflow/models/dag.py:180]: Oh boy I did not realize this was a use case

File: airflow/www/security.py (hunk 13 above)
Comment (jhtimmins) [reply airflow/www/security.py]: Ohhhhhh I just realized that this solves a bug in a different PR

File: airflow/api_connexion/security.py (hunk 10 above)
Comment (ashb) [reply airflow/api_connexion/security.py]: I'd be happier with 'Dags' than 'AllDags', yes.

File: airflow/api_connexion/endpoints/task_instance_endpoint.py [hunk 15]
```diff
@@ -43,7 +43,7 @@
 
 @security.requires_access(
     [
-        ("can_read", "Dag"),
+        ("can_read", "AllDags"),
```
Comment (ashb) [inline airflow/api_connexion/endpoints/task_instance_endpoint.py]: Why is this one AllDags, not Resource.DAGS?

Comment (jhtimmins) [reply airflow/api_connexion/endpoints/task_instance_endpoint.py]: Ah this is because this endpoint is new. Updated.

File: airflow/models/dag.py [hunk 16]
```diff
@@ -88,6 +89,23 @@ def get_last_dagrun(dag_id, session, include_externally_triggered=False):
     return query.first()
 
 
+def replace_outdated_dag_actions(access_control):
```
Comment (ashb) [inline airflow/models/dag.py]: I think make this an `@staticmethod` on dag, and then also call it `_upgrade_outdated_dag_access_control`

File: airflow/models/dag.py [hunk 17]
```diff
@@ -88,6 +89,23 @@ def get_last_dagrun(dag_id, session, include_externally_triggered=False):
...
+    access_controls (for example, {'role1': {'can_dag_read'}, 'role2': {'can_dag_read', 'can_dag_edit'}})
+    and replaces them with updated permissions (can_read and can_edit).
+    """
+    new_perm_mapping = {
+        permissions.DEPRECATED_ACTION_CAN_DAG_READ: permissions.ACTION_CAN_READ,
+        permissions.DEPRECATED_ACTION_CAN_DAG_EDIT: permissions.ACTION_CAN_EDIT,
+    }
+    updated_access_control = {}
+    for role, perms in access_control.items():
+        updated_access_control[role] = {new_perm_mapping.get(perm, perm) for perm in perms}
```
Comment (ashb) [inline airflow/models/dag.py]: Could issue a DeprecationWarning if anything was changed? (with the right `stacklevel` to have the line number appear where the `DAG()` call is, 3 I think?)

File: tests/api_connexion/endpoints/test_task_instance_endpoint.py [hunk 18]
```diff
@@ -46,7 +46,7 @@ def setUpClass(cls) -> None:
             username="test",
             role_name="Test",
             permissions=[
-                ('can_read', 'Dag'),
+                ('can_read', 'AllDags'),
```
Comment (ashb) [inline tests/api_connexion/endpoints/test_task_instance_endpoint.py]: Needs updating.

File: tests/test_utils/asserts.py [hunk 19]
```diff
@@ -74,11 +77,16 @@ def assert_queries_count(expected_count, message_fmt=None):
...
-        message_fmt = message_fmt or "The expected number of db queries is {expected_count}. " \
-                                     "The current number is {current_count}.\n\n" \
-                                     "Recorded query locations:"
-        message = message_fmt.format(current_count=count, expected_count=expected_count)
+    if count != expected_count:
+        message_fmt = (
+            message_fmt
+            or "The expected number of db queries is {expected_count}. "
+            "The current number is {current_count}.\n\n"
+            "Recorded query locations:"
```
Comment (ashb) [inline tests/test_utils/asserts.py]: Could you revert the changes to this file please (since this is now just whitespace only in this file)

**Review by ashb — CHANGES_REQUESTED** [review]
Getting there!

Could you also add a note in UPDATING.md about the change of permissions and the change to the `access_control` param.

**Review by ashb — APPROVED** [review]

**Comment by kaxil** [comment]
🚀

**Comment by jhtimmins** [comment]
Oh dang, thanks @kaxil !
